[LLDB] Add type casting to DIL, part 1 of 3.

cmtice opened the pull request:
This is an alternative to https://github.com/llvm/llvm-project/pull/159500, breaking that PR down into three separate PRs, to make it easier to review.

This first PR of the three adds the basic framework for doing type casing to the DIL code, but it does not actually do any casting: In this PR the DIL parser only recognizes builtin type names, and the DIL interpreter does not do anything except return the original operand (no casting). The second and third PRs will add most of the type parsing, and do the actual type casting, respectively.

cmtice committed:
[LLDB] Add type casting to DIL, part 1 of 3.

This is an alternative to PR 159500, breaking that PR down
into three separate PRs, to make it easier to review.

This first PR of the three adds the basic framework for doing
type casing to the DIL code, but it does not actually do any
casting: In this PR the DIL parser only recognizes builtin type
names, and the DIL interpreter does not do anything except
return the original operand (no casting). The second and third PRs
will add most of the type parsing, and do the actual type casting,
respectively.

Michael137 commented on lldb/source/ValueObject/DILParser.cpp:
```diff
@@ -80,15 +82,62 @@ ASTNodeUP DILParser::Run() {
+
+ASTNodeUP DILParser::ParseCastExpression() {
+  // This can be a C-style cast, try parsing the contents as a type declaration.
+  if (CurToken().Is(Token::l_paren)) {
```
Can we do use early-return style here?

```suggestion
  if (!CurToken().Is(Token::l_paren))
    return ParseUnaryExpression();
    
  ...
```

Michael137 commented on lldb/source/ValueObject/DILParser.cpp:
```diff
@@ -339,6 +463,40 @@ std::string DILParser::ParseUnqualifiedId() {
+                              type.TypeDescription()),
+                loc, CurToken().GetSpelling().length());
+        return bad_type;
+      }
```
Makes me think, should we support references at all for now? Might be wrong, but I'd think casting to a reference type is not a super common use-case. Once we add `&` as a `ptr_operator`, we're saying `&` is the reference syntax for all languages.

Again, if this has already been discussed elsewhere, feel free to ignore

Michael137 replied:
> But in general, this is a C-style cast with C-style syntax, I think it should be expected to be used only with C/C++ code

But isn't DIL supposed to be the de-facto inspection language? Regardless of what language you're debugging. Yes it follows mostly C/C++ syntax but here you are checking explicitly C++ language semantics, not just syntax.

Maybe I'm incorrectly delineating DIL and the language plugins.

E.g., an alternative would be to say:
```
if (tk.GetKind() == Token::star)
  type = type.GetPointerType();
  if (!type) {
     BailOut("couldn't get a pointer type to ....");
     return {};
  }
```
And the language plugin decide whether it can create a pointer to a reference type.

Wdyt?

jimingham replied:
Since the DIL is interpreting:  `(Typename <operator>) Identifier` to mean "re-interpret Identifier as `an <operator> variant of "Typename"` regardless of how the underlying language would actually spell that operation, similarly `*` and `&` can't just mean "pass a * to the language and see what it does with it" since the underlying language may very well not spell it that way.
So if the DIL can convince itself some combination of primitive operations in a cast doesn't make sense in the abstract, it should be free to reject it.  But if it just isn't allowed when spelled that way in the languages we know, then it should pass that off to the type system.

kuilpd commented on lldb/source/ValueObject/DILParser.cpp:
```diff
@@ -274,6 +323,81 @@ std::string DILParser::ParseNestedNameSpecifier() {
+  bool first_word = true;
+  while (CurToken().GetKind() == Token::identifier) {
+    if (CurToken().GetSpelling() == "const" ||
+        CurToken().GetSpelling() == "volatile")
```
From what I understand, only `const_cast` can be used to cast to cv-qualifiers. The compiler also drops them when using a C-style cast.

cmtice replied:
We are not planning on processing or doing anything with the CV qualifiers. We *could* give a  parsing error here, but it seemed kinder to just ignore them?

cmtice replied:
Actually, I take that back.  We can't exactly give error messages on them, because we don't know (at this point during the parse) if we're actually handling a type name or not. It's conceivable that this could be an expression in a language where "const" or "volatile" are legal identifier names. Or it could really be a type name and the user threw these in as qualifiers...but we're not handling the CV qualifiers as part of the cast....So silently dropping them seems the best thing we can do here. Or I suppose if we enounter them we could treat it as definitely NOT a type name, but that seems weird, too.

cmtice commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -3,10 +3,13 @@
 
 unary_expression = postfix_expression
-                 | unary_operator expression ;
+                 | unary_operator cast_expression ;
```
In the currently shown grammar yes. But the full grammar for the full DIL expression language has quite a few layers between "expression" and "cast_expression", and I don't think it would be correct to allow some of them here. So I would prefer to keep this as is at the moment, in anticipation of future changes. But I can make your suggested change if you feel strongly about it.

cmtice commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -35,6 +36,14 @@ enum class UnaryOpKind {
   Deref,  // "*"
 };
 
+/// The C-Style casts allowed by DIL.
```
Actually, no I think we reached an agreement not to add those other types of casts to DIL, because they were too language-specific.  These comments referring explicitly to C-Style casts are from the implementation that allowed for the other C++ styles of casting, but since we're not planning on adding that to DIL, I will remove the "C Style" references.

cmtice commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -35,6 +36,14 @@ enum class UnaryOpKind {
+  eEnumeration,
+  eNullptr,
+  eReference,
+  eNone,
```
Done.

cmtice committed:
Address reviewer comments:
	- change references to C-style casting to just casting
	- add comments to cast enumeration types
	- use early-return

jimingham commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -608,4 +608,16 @@ Interpreter::Visit(const BooleanLiteralNode *node) {
+  lldb::ValueObjectSP operand = *operand_or_err;
+  // Don't actually do the cast for now -- that code will be added later.
+  // For now just return the original operand, unchanged.
+  return operand;
```
I'm worried about this terminology.  `frame var` is not and should not be expression evaluation.  It's for static inspection of memory, it doesn't do operator overloads in general, and doesn't construct temporary objects or call copy constructors or anything like that.  It just looks at static values in memory.  We should not give the impression that `frame var` is the expression evaluator.

This project started because some folks wanted to make it easier to write data formatters (which should be non-code-running for efficiency & stability) using "value path expressions" that required some operations we didn't support in the `frame var` language.  In that use case, presumably the authors know what they are doing and won't confuse DIL with the real language specific expression evaluator.

Even `dwim-print` requires context in this regard.  After all, an -> access of a value that has an -> overload is going to be different in dwim-print.  That command works for common cases but really does require the user to know what they are doing, since there will be cases where both `frame var` and `expr` return valid but different results.
So it seems to me we should be careful of anything we say or do that confuses the two.

cmtice replied:
Returning (no pun intended) to the original question here: Is it OK, for now, for this function to return the unchanged operand? Especially because the PR to do the actual type casting will be sent for review as soon as this one lands?

cmtice replied:
If offline conversation kuilpd convinced me that this should return an error for now, since it's not actually doing the typecast. Have updated the code to do that.

cmtice committed:
Rmove unnecessary 'bad_type'; add FIXME comment.

cmtice committed:
Fix doxygen comments.

cmtice commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -35,6 +36,14 @@ enum class UnaryOpKind {
 
+/// The type casts allowed by DIL.
+enum class CastKind {
+  eEnumeration, /// Casting from a scalar to an enumeration type
```
Oh! I didn't know that. Thanks!  (Fixed).

cmtice committed:
Merge remote-tracking branch 'origin/main' into dil-typecast-outline

cmtice committed:
Update DIL interpreter to return an error message for type casting.

Will replace this with code to do actual casts in next PR.

cmtice committed:
Fix clang format issues.

adrian-prantl reviewed (approved):
My only reservation to this path is that the syntax itself is very C-specific, but there no reason why we could not at a later point in time allow for language plugins to define their own DIL dialects that use different parsers to create a valid DILAST. 

Nothing in this patch prevents us from doing that later on.
